lib/ecto/migration.ex

```diff
@@ -984,17 +1000,49 @@ defmodule Ecto.Migration do
   defp validate_index_opts!(opts), do: opts
 
   @doc false
-  def __prefix__(%{prefix: prefix} = index_or_table) do
```

Comment:
This `index_or_table` argument is what prompted the thought that we might want to pattern match to be a bit more stringent about what type of structs these functions are operating on. The test suite passes `%Constraint{}` structs into this `__prefix__/1` function